### PR TITLE
HHH-16185 Standardize the date_trunc() function, emulating it for Dialects that don't support native versions

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-16185

### CODE REVIEW

**Comment by gavinking** [comment]
So the idea is format it as a string and then parse the string?

**Comment by mbladel** [comment]
> So the idea is format it as a string and then parse the string?

Yep, when no native way of truncating exists.

**Comment by gavinking** [comment]
Sounds good. (And useful.) But we need to figure out what we're going to call this function, since `date_trunc` doesn't conform to our conventions.

**Comment by mbladel** [comment]
> Sounds good. (And useful.) But we need to figure out what we're going to call this function, since `date_trunc` doesn't conform to our conventions.

Sure, it should be easy enough to change the name. The convention is all lower case and no underscores, correct?

So maybe `datetrunc`?

**Comment by gavinking** [comment]
> So maybe `datetrunc`?

Is it only for dates? Isn't it also for timestamps?

**Comment by mbladel** [comment]
> Is it only for dates? Isn't it also for timestamps?

Yes it is. It should support all `TEMPORAL` types.

**Comment by gavinking** [comment]
So the _natural_ thing to write here would be something like:

    truncate(datetime, second)

or even:

    truncate(datetime to second)

but of course this would mean overloading the existing `truncate` function. Not sure how I feel about that.

The reason I suggest the second option is that `to` is already a keyword, and we already support the ANSI syntax `extract(second from datetime)`.

**Comment by sebersole** [comment]
I think the form `truncate(datetime, second)` is reasonable; it is consistent with the current 2-arg format.  The form using `to` changes the expected signature structure.  The overloading it natural imo.

**Comment by mbladel** [comment]
While trying to rename the function to override the numeric `truncate` I noticed we might have a problem with the `truncate(expression, temporal_unit)` syntax.

If we have to disambiguate between the numeric and the datetime internal implementation, how can we know if the second parameter is being used as a `temporal_unit` (second, hour, minute) or if it's a path expression for the second parameter of the numeric `trunc` function?

The syntax `truncate(expression to temporal_unit)` could let us know for sure which type of function the user is calling. Otherwise a dedicated function name for datetime truncation would work as well.

**Comment by gavinking** [comment]
Yeah we don’t currently have any infrastructure to support function overloading. We would have to build that.

**Comment by sebersole** [comment]
So first, after looking at `TemporalUnit`, we actually have values there that are not valid for truncate (`WEEK_OF_MONTH`, `WEEK_OF_YEAR`,...).  That enum was developed for `extract`, so this is fine-ish.  In retrospect, I'd say these are more "temporal extraction units", but that's a minor naming concern.  But it is important to understand in relation to this discussion.  Here we want to allow truncation to actual temporal units, though honestly some "real" probably don't make a lot of practical sense - NANOSECOND, SECOND and even YEAR.  With that said...

> If we have to disambiguate between the numeric and the datetime internal implementation, how can we know if the second parameter is being used as a `temporal_unit` (second, hour, minute) or if it's a path expression for the second parameter of the numeric `trunc` function?

Given that we'd have a hard-coded list of values that we handle as a temporal-unit here, I'd just recognize them specifically - either 

- as keywords, if these are defined that way already, 
- as "soft" or "contextual" keywords

Also, FWIW, I'm not sure a path expression for the "scale" argument is very useful.  We could chose to limit the "scale" expression to just literals and parameters.  That would help with the "is it  a path" aspect - it would never be.


> The syntax `truncate(expression to temporal_unit)` could let us know for sure which type of function the user is calling. Otherwise a dedicated function name for datetime truncation would work as well.

If we end up having to go this route I am not a fan of simply reusing the name "truncate".  I think this suggestion goes well beyond "overloading" as we all understand it.

**Comment by gavinking** [comment]
> Given that we'd have a hard-coded list of values that we handle as a temporal-unit here, I'd just recognize them specifically

Yes, that's the right way.

**Comment by mbladel** [comment]
FYI updated the PR with:
* custom implementation for `trunc()` which handles both the numeric and datetime variant
* new parser rule for "contextual" temporal units
* updated UG with the new `trunc()` function for datetimes

**Review by beikov — CHANGES_REQUESTED** [review]
Great work. I left a few comments, but overall, this looks nice!

File: hibernate-core/src/main/java/org/hibernate/dialect/function/CommonFunctionFactory.java [hunk 1]
```diff
@@ -356,8 +350,8 @@ public void truncate() {
 	public void truncate_round() {
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/function/CommonFunctionFactory.java:350]: This was not used, removed the method from function factory.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/CommonFunctionFactory.java [hunk 2]
```diff
@@ -259,93 +259,87 @@ public void moreHyperbolic() {
...
-				"sign(?1)*floor(abs(?1))",
-				"sign(?1)*floor(abs(?1)*1e?2)/1e?2",
-				NUMERIC, INTEGER,
-				typeConfiguration
-		).setArgumentListSignature( "(NUMERIC number[, INTEGER places])" );
-		functionRegistry.registerAlternateKey( "truncate", "trunc" );
+		trunc( "sign(?1)*floor(abs(?1))", "sign(?1)*floor(abs(?1)*1e?2)/1e?2", null, null );
 	}
 
 	public void truncate() {
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/function/CommonFunctionFactory.java]: This was used in a couple community dialects, I now substituted it with `trunc_truncate()` and removed the unneeded method from the factory.

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/hql/FunctionTests.java [hunk 3]
```diff
@@ -512,13 +512,17 @@ public void testRoundTruncFunctions(SessionFactoryScope scope) {
...
 	public void testDateTruncFunction(SessionFactoryScope scope) {
 		scope.inTransaction(
-				session -> session.createQuery("select date_trunc(year,current_timestamp)", Timestamp.class).getSingleResult()
+				session -> {
+					session.createQuery( "select trunc(current_timestamp,year)", Timestamp.class ).getSingleResult();
+					session.createQuery( "select trunc(current_timestamp,month)", Timestamp.class ).getSingleResult();
+					session.createQuery( "select trunc(current_timestamp,day)", Timestamp.class ).getSingleResult();
+					session.createQuery( "select truncate(current_timestamp,hour)", Timestamp.class ).getSingleResult();
+					session.createQuery( "select truncate(current_timestamp,minute)", Timestamp.class ).getSingleResult();
+					session.createQuery( "select truncate(current_timestamp,second)", Timestamp.class ).getSingleResult();
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/test/java/org/hibernate/orm/test/query/hql/FunctionTests.java:524]: Yes, thanks to the `(TRUNC | TRUNCATE)` parser rule. It should also be tested by `org.hibernate.orm.test.query.hql.FunctionTests#testRoundTruncFunctions`.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/TruncFunction.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+			}
+			if ( datetimeTrunc.getPattern() == null ) {
+				final boolean useConvertToFormat = nodeBuilder.getSessionFactory()
+						.getJdbcServices()
+						.getDialect() instanceof SybaseDialect;
+				return new DateTruncEmulation(
+						toDateFunction,
+						useConvertToFormat,
+						typeConfiguration
+				).generateSqmFunctionExpression( arguments, impliedResultType, queryEngine, typeConfiguration );
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/function/TruncFunction.java]: You're right, somehow I thought I saw some dialect-checking logic in custom functions but I was wrong.

I've now created dedicated functions to handle dialect-specific cases (Sybase, Oracle, PostgreSQL and SQLServer < 16).

File: hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+			// see: https://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.infocenter.dc36271.1600/doc/html/san1393050437990.html
+			sqlAppender.append( "datetime,substring(convert(varchar," );
+			sqlAstArguments.get( 0 ).accept( walker );
+			sqlAppender.append( ",21),1,17-len(" );
+			sqlAstArguments.get( 1 ).accept( walker );
+			sqlAppender.append( "))+" );
+			sqlAstArguments.get( 1  ).accept( walker );
+			sqlAppender.append( ",21" );
+		}
+		sqlAppender.append( ')' );
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java]: Created additional `DateTruncConvertEmulation` and `SybaseDateTruncEmulation` that extend the base emulation function for specific cases. `DateTruncEmulation` is still used to handle "regular" emulation.

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/hql/FunctionTests.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/test/java/org/hibernate/orm/test/query/hql/FunctionTests.java:524]: Ok, but using the `truncate` function in e.g. Criteria wouldn't work this way anymore. Can you please add a test for that and re-add the alternate keys?

File: hibernate-core/src/main/java/org/hibernate/dialect/function/TruncFunction.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+ * @author Marco Belladelli
+ */
+public class TruncFunction extends AbstractSqmFunctionDescriptor implements FunctionRenderingSupport {
+	private final PatternRenderer truncPattern;
+	private final PatternRenderer twoArgTruncPattern;
+	protected final DatetimeTrunc datetimeTrunc;
+	private final PatternRenderer datetimeTruncPattern;
+	private final String toDateFunction;
+
+	protected boolean numericTrunc;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/TruncFunction.java:49]: That's not going to work. The function descriptor is shared across everything, so having state here is not possible. You can instead create two implementations of `FunctionRenderingSupport` and use one or the other when instantiating `SelfRenderingSqmFunction`

File: hibernate-core/src/main/java/org/hibernate/dialect/function/SybaseTruncFunction.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	@Override
+	protected <T> SelfRenderingSqmFunction<T> generateSqmFunctionExpression(
+			List<? extends SqmTypedNode<?>> arguments,
+			ReturnableType<T> impliedResultType,
+			QueryEngine queryEngine,
+			TypeConfiguration typeConfiguration) {
+		final List<SqmTypedNode<?>> args = new ArrayList<>( arguments );
+		if ( arguments.size() == 2 && arguments.get( 1 ) instanceof SqmExtractUnit ) {
+			// datetime truncation
+			return new SybaseDateTruncEmulation( "convert", typeConfiguration ).generateSqmFunctionExpression(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/SybaseTruncFunction.java:43]: Would be nice if you could keep the `SybaseDateTruncEmulation` in a private final variable initialized in the constructor.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/SqlServerConvertTruncFunction.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	@Override
+	protected <T> SelfRenderingSqmFunction<T> generateSqmFunctionExpression(
+			List<? extends SqmTypedNode<?>> arguments,
+			ReturnableType<T> impliedResultType,
+			QueryEngine queryEngine,
+			TypeConfiguration typeConfiguration) {
+		final List<SqmTypedNode<?>> args = new ArrayList<>( arguments );
+		if ( arguments.size() == 2 && arguments.get( 1 ) instanceof SqmExtractUnit ) {
+			// datetime truncation
+			return new DateTruncConvertEmulation( typeConfiguration ).generateSqmExpression(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/SqlServerConvertTruncFunction.java:43]: Would be nice if you could keep the `DateTruncConvertEmulation` in a private final variable initialized in the constructor.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/PostgreSQLTruncFunction.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+			QueryEngine queryEngine,
+			TypeConfiguration typeConfiguration) {
+		final List<SqmTypedNode<?>> args = new ArrayList<>( arguments );
+		if ( arguments.size() == 2 && arguments.get( 1 ) instanceof SqmExtractUnit ) {
+			// datetime truncation
+			numericTrunc = false;
+		}
+		else {
+			// numeric truncation
+			return new PostgreSQLTruncRoundFunction( "trunc", supportsTwoArguments ).generateSqmFunctionExpression(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/PostgreSQLTruncFunction.java:50]: Would be nice if you could keep the `PostgreSQLTruncRoundFunction` in a private final variable initialized in the constructor.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/OracleTruncFunction.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+					break;
+				case HOUR:
+					pattern = "HH";
+					break;
+				case MINUTE:
+					pattern = "MI";
+					break;
+				case SECOND:
+					// Oracle does not support truncating to seconds with the native function, use emulation
+					return new DateTruncEmulation( "to_date", typeConfiguration )
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/OracleTruncFunction.java:72]: Would be nice if you could keep the `DateTruncEmulation` in a private final variable initialized in the constructor.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+		final SqmFormat sqmFormat = new SqmFormat(
+				"yyyy-MM-dd HH:mm:ss",
+				typeConfiguration.getBasicTypeForJavaType( String.class ),
+				nodeBuilder
+		);
+
+		return new SelfRenderingSqmFunction<>(
+				this,
+				this,
+				asList( datetime, formattedDatetime, sqmFormat ),
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java:152]: Seems the first argument is only used in `SybaseDateTruncEmulation#render` method. Can you please add a comment here about this?

File: hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+							),
+							null,
+							queryEngine,
+							typeConfiguration
+					);
+		}
+		else {
+			formattedDatetime = formatExpression;
+		}
+		final SqmFormat sqmFormat = new SqmFormat(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java:143]: Would be nice if you could keep the `SqmFormat` in a private final variable initialized in the constructor.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/TruncFunction.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+			// numeric truncation
+			numericTrunc = true;
+		}
+
+		return new SelfRenderingSqmFunction<>(
+				this,
+				this,
+				args,
+				impliedResultType,
+				getArgumentsValidator(),
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/TruncFunction.java:178]:
```suggestion
				numericTrunc ? TruncArgumentsValidator.NUMERIC_VALIDATOR : TruncArgumentsValidator.DATETIME_VALIDATOR,
```

File: hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java (hunk 12 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java:143]: If I'm initializing `DateTruncEmulation` as a private final variable in the constructor of `TruncFunction` I don't think I can access the `NodeBuilder` to pass to `SqmFormat` in the constructor.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/function/DateTruncEmulation.java:143]: The node builder is actually not important here. You could pass null for it.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/SybaseTruncFunction.java (hunk 7 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/function/SybaseTruncFunction.java:43]: Should I also make dialect-specific datetime trunc emulation functions inner classes of the already dialect-specific trunc functions (e.g. `SybaseTruncFunction$SybaseDateTruncEmulation`)?

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/function/SybaseTruncFunction.java:43]: Don't know if I understand exactly, but whatever you think works best ;)

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/hql/FunctionTests.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/test/java/org/hibernate/orm/test/query/hql/FunctionTests.java:525]: Oh my bad, added the test in `CriteriaBuilderNonStandardFunctionsTest`. Also applied all latest suggestions.

**Review by beikov — APPROVED** [review]
LGTM

**Comment by ManjunathM22** [comment]
> https://hibernate.atlassian.net/browse/HHH-16185

HI ,,

I have been badly stuck up with - date_trunc function can please help me -- my code is as below

Line - 1

private Date slotSystem;
timeFormat - yyyy-mm-dd

to_char(date_trunc('%1$s', c.slotSystem), :timeFormat) as period, " 

Line - 2
"     datetrunc('%1$s', c.slotSystem) "

Exception ; 
org.postgresql.util.PSQLException: ERROR: function datetrunc(unknown, bigint) does not exist
  Hint: No function matches the given name and argument types. You might need to add explicit type casts.

**Comment by gavinking** [comment]
@ManjunathM22 The function is now called `truncate()`, as documented here:

<https://docs.jboss.org/hibernate/orm/6.2/userguide/html_single/Hibernate_User_Guide.html#hql-functions-datetime>

**Pull request merged**